README.md

# SNPgrou

<!-- badges: start -->
<!-- badges: end -->

The goal of SNPgrou is to associate multi-level phenotypic traits with multiple loci to genome wide SNP data. 

## Installation

You can install the released version of SNPgrou by:

``` r
require("devtools")
install_github("Kevinpen/SNPgrou", build_vignettes = TRUE)
library("SNPgrou")
```
To run shinyApp:

``` r
runSNPgrou()
```


## Overview

This R package is intended to use as a tool to assist GWAS: asscociate phenotypic response to genotype data. In existing GWAS packages, phenotype response variable are limited to continues or binary, this package specifically extends the response to categorical and have multiple levels. So this package can be used for situations where phenotype data are like eye colours or subtypes of breast cancer. 

This package use lda function from MASS package, but to give an explicit metric of how each locus is related to the phenotype, it build a "gScore", which is the measure of how accurately the lda model generate phenotype in repeated cross-validation. This practice is based on the idea that when using leave-one-out cross-validation to predicate the phenotype, those locus with strong relation to phenotype should be able to accurately predicate phenotype. 

An overview of the package is illustrated below. 

![](./inst/extdata/PENG_G_A1.png) 



## Contributions

The author of the package is Gang Peng. The functions available within this package include:

``` r
library("SNPgrou")
lsf.str("package:SNPgrou")
```
- prep
- snpgrou
- gws
- groupView
- plotGws
- summaryGws
- is.gScore
- summarySNP
- is.SNP

The package use GWS data from package "SNPassoc", but the original phenotype data is binary, here three levels phenotype is simulated. The workflow of this package is simialr to SNPassoc. The concept gSCore is similar to  "Classification accuracy" in the package "genephen", which use other statistical learning model.

## References 
- González JR1, Armengol L, Solé X, Guinó E, Mercader JM, Estivill X, Moreno V(2007). SNPassoc: an R package to perform whole genome association studies. Bioinformatics. 2007 Mar 1;23(5):644-5. Epub 2007 Jan 31. DOI: 10.1093/bioinformatics/btm025.

- Thanks package "genphen"(DOI: 10.18129/B9.bioc.genphen, Author: Simo Kitanovski) inspired the concept of gScore. 

## Acknowledgments

- This package was developed as part of an assessment for 2019 BCB410H: Applied Bioinformatics, University of Toronto, Toronto, CANADA.




NEWS.md
# SNPgrou 0.1.0

* Added a `NEWS.md` file to track changes to the package.
* Added summaryGWS and summarySNP function


R/prep.R
#' Prepare the data for the analysis purpose
#'
#' Transform the data to class "SNP" object, with alleles separated with "/"
#'
#' @param x, Factors, data of all obeservation of alleles in single locus
#' @param delim character that separate the alleles in original data in one locus, could be empty
#'
#'
#' @return object belong to "SNP" and "factor", with attributes of alleleNames
#'
#' @examples
#' data(geneSNP)
#' s1 <- prep(geneSNP$snp10001,delim="-") # This column original delimiter "-", replaced by "/"
#' s2 <- prep(geneSNP$snp10002,delim=",") # This column original delimiter "," replaced by "/"
#' s3 <- prep(geneSNP$snp10003)           # This column original data no delimiter, add "/"
#'
#'@references
#'González JR1, Armengol L, Solé X, Guinó E, Mercader JM, Estivill X, Moreno V(2007).
#'SNPassoc: an R package to perform whole genome association studies. Bioinformatics. 2007
#'Mar 1;23(5):644-5. Epub 2007 Jan 31. DOI: 10.1093/bioinformatics/btm025.
#'
#' @export

prep <-
  function(x, delim = NULL) {
    if (is.null(x)) {
      stop("No input data found.")
    }
    x <- as.character(x)

    # Legitimate SNP data format conform to the convention(like in package "SNPassoc" and "GenABEL")
    # of 2 alleles separated by delimiter, after preparation delimiter changed to "/"

    # Check whether the SNP data is legitimate, for no delimiter case
    if (all(lapply(x, nchar) == 2, na.rm = TRUE)) {
      if (!(all(grepl("[ACGT]", x), na.rm = TRUE))) {
        stop(" Data contain non SNP data")
      }
    }

    # Check whether the SNP data is legitimate, for delimiter present case
    if (all(lapply(x, nchar) == 3, na.rm = TRUE)) {
      y <- c(substring(x, 1, 1), substring(x, 3, 3))
      if (!(all(grepl("[ACGT]", stats::na.omit(y))))) {
        stop(" Data contain non SNP data")
      }
    }

    # Legitmate SNP data for single locus can only be 2 or 3 characters
    else if (all(lapply(x, nchar) > 3, na.rm = TRUE)) {
      stop("SNP data exceed 3 character limit, or less than 2 character")
    }

    # No delim parameter specified
    if (missing(delim)) {
      # No delim parameter specified and the actual data do not contain delimiter, add delimiter"/"
      if (all(lapply(x, nchar) == 2, na.rm = TRUE)) {
        subStr <- paste(substring(x, 1, 1), substring(x, 2, 2), sep = "/")
        alleles <-
          unique(stats::na.omit(c(
            substring(x, 1, 1), substring(x, 2, 2)
          )))

        # No delim parameter specified, but actual data contain delimiter, substitue delimiter to "/"
      } else {
        subStr <- paste(substring(x, 1, 1), substring(x, 3, 3), sep = "/")
        alleles <-
          unique(stats::na.omit(c(
            substring(x, 1, 1), substring(x, 3, 3)
          )))
      }

      # Delimiter specified as ""
    } else if (delim == "") {
      subStr <- paste(substring(x, 1, 1), substring(x, 2, 2), sep = "/")
      alleles <-
        unique(stats::na.omit(c(substring(x, 1, 1), substring(x, 2, 2))))

    } else if (is.character(delim)) {
      # Test whether the user specified delimiter contained in the data
      if (!all(grepl(delim, stats::na.omit(as.character(x)), fixed = TRUE))) {
        stop("The delimiter you specified does not contained in your data")

      } else {
        # Change original deilimiter to "/"
        subStr <- gsub(delim, "/", x)
        alleles <-
          unique(stats::na.omit(c(
            substring(x, 1, 1), substring(x, 3, 3)
          )))
      }

    } else {
      # The dilim parameter is not legitimate delimiter
      stop("Please use the correct dilimiter in the data")
    }

    mode(subStr) <- "character"
    result <- factor(subStr)

    #create SNP class
    class(result) <- c("SNP", "factor")
    attr(result, "alleleNames") <- alleles

    return(result)
  }
# [END]

